src/Simplify_Call.cpp

```diff
@@ -0,0 +1,417 @@
+                    Expr new_extent = mutate(extent_0 * extent_1, nullptr);
+                    Expr new_stride = stride_0;
+                    args.erase(args.begin() + j, args.begin() + j + 2);
+                    args[i] = new_extent;
+                    args[i + 1] = new_stride;
+                    i -= 2;
+                    break;
+                }
+            }
+        }
+        internal_assert(args.size() <= op->args.size());
+
+        if (changed || (args.size() != op->args.size())) {
+            return Call::make(op->type, Call::prefetch, args, Call::Intrinsic);
+        } else {
+            return op;
+        }
+    } else if (op->is_intrinsic(Call::require)) {
+        Expr cond = mutate(op->args[0], nullptr);
+        // likely(const-bool) is deliberately not reduced
+        // by the simplify(), but for our purposes here, we want
+        // to ignore the likely() wrapper. (Note that this is
+        // equivalent to calling can_prove() without needing to
+        // create a new Simplifier instance.)
+        if (const Call *c = cond.as<Call>()) {
+            if (c->is_intrinsic(Call::likely)) {
+                cond = c->args[0];
+            }
+        }
+
+        if (is_zero(cond)) {
+            // (We could simplify this to avoid evaluating the provably-false
+            // expression, but since this is a degenerate condition, don't bother.)
+            user_warning << "This pipeline is guaranteed to fail a require() expression at runtime: \n"
+                         << Expr(op) << "\n";
+        }
+
+        Expr result;
+        {
+            // Can assume the condition is true when evaluating the value.
+            auto t = scoped_truth(cond);
+            result = mutate(op->args[1], bounds);
+        }
+
+        if (is_one(cond)) {
+            return result;
+        }
+
+        Expr message = mutate(op->args[2], nullptr);
+
+        if (cond.same_as(op->args[0]) &&
+            result.same_as(op->args[1]) &&
+            message.same_as(op->args[2])) {
+            return op;
+        } else {
+            return Internal::Call::make(op->type,
+                                        Internal::Call::require,
+                                        {std::move(cond), std::move(result), std::move(message)},
+                                        Internal::Call::PureIntrinsic);
+        }
+    } else {
+        vector<Expr> new_args(op->args.size());
+        bool changed = false;
+
+        // Mutate the args
+        for (size_t i = 0; i < op->args.size(); i++) {
+            const Expr &old_arg = op->args[i];
+            Expr new_arg = mutate(old_arg, nullptr);
+            if (!new_arg.same_as(old_arg)) changed = true;
+            new_args[i] = std::move(new_arg);
+        }
+
+        if (!changed) {
+            return op;
+        } else {
+            return Call::make(op->type, op->name, new_args, op->call_type,
```

Comment:
I inserted something like this here, to catch provable divmod-by-zero earlier:
```
            if (new_args.size() == 2 && (op->is_intrinsic(Call::quiet_div) || op->is_intrinsic(Call::quiet_mod)) && is_zero(new_args[1])) {
              user_error << "Attempting to divmod by constant zero:\n"
                << "    numer "<<op->args[0]<<" -> "<<new_args[0]<<"\n"
                << "    denom "<<op->args[1]<<" -> "<<new_args[1]<<"\n\n";
            }
```

Reply:
But that should explicitly be legal! The whole point of quiet_div is that it's legal (well, side-effect-free) to evaluate it with a zero denominator, you just get an undefined value, so it had better be masked off by a select higher up in the AST.

Reply:
(of course, I suppose these are the ones we want to let thru, so the change of this sort I really want is in the handle for Div andMod...)

Reply:
Yeah, see followup comment -- realized after commenting :-)